source/server/listener_impl.cc

```diff
@@ -1123,6 +1145,39 @@ void ListenerImpl::closeAllSockets() {
   }
 }
 
+bool ListenerMessageUtil::compareSocketOptions(const envoy::config::listener::v3::Listener& lhs,
+                                               const envoy::config::listener::v3::Listener& rhs) {
+  if ((PROTOBUF_GET_WRAPPED_OR_DEFAULT(lhs, transparent, false) !=
+       PROTOBUF_GET_WRAPPED_OR_DEFAULT(rhs, transparent, false)) ||
+      (PROTOBUF_GET_WRAPPED_OR_DEFAULT(lhs, freebind, false) !=
+       PROTOBUF_GET_WRAPPED_OR_DEFAULT(rhs, freebind, false)) ||
+      (PROTOBUF_GET_WRAPPED_OR_DEFAULT(lhs, tcp_fast_open_queue_length, 0) !=
+       PROTOBUF_GET_WRAPPED_OR_DEFAULT(rhs, tcp_fast_open_queue_length, 0))) {
+    return false;
+  }
+
+  if (lhs.socket_options_size() != rhs.socket_options_size()) {
+    return false;
+  }
+
+  for (auto i = 0; i < lhs.socket_options_size(); i++) {
+    auto& option = lhs.socket_options(i);
+    bool found = false;
+    for (auto j = 0; j < rhs.socket_options_size(); j++) {
+      auto& other_option = rhs.socket_options(j);
+      Protobuf::util::MessageDifferencer differencer;
+      if (differencer.Compare(option, other_option)) {
+        found = true;
+        break;
+      }
+    }
+    if (!found) {
+      return false;
+    }
+  }
+  return true;
+}
```

Comment:
I don't feel very strongly about it, but you could probably make this simpler by just doing a direct comparison vs. an unordered comparison. Since it's an array the order shouldn't change and I don't personally think it's worth checking for unordered equal but up to you.

Reply:
+1. I'd rather keep it simpler.